[ERROR] Flutter desktop macOS: Storage error "Operation not permitted" (code 1)

- ObjectBox version: 1.0.0
- Flutter/Dart SDK: 2.0.6
- Null-safety enabled: yes
- Reproducibility: always
```
flutter run

Running "flutter pub get" in untitled...                         1,220ms
Launching lib/main.dart on macOS in debug mode...
Running pod install...                                           1,395ms
Building macOS application...                                           
001-10:23:16.4099 [ERROR] Storage error "Operation not permitted" (code 1)
[ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: ObjectBoxException: failed to create store: 10199 Could not open database environment; please check options and file system (1: Operation not permitted)
#0      ObjectBoxNativeError.throwMapped (package:objectbox/src/native/bindings/helpers.dart:78:9)
#1      throwLatestNativeError (package:objectbox/src/native/bindings/helpers.dart:50:48)
#2      checkObxPtr (package:objectbox/src/native/bindings/helpers.dart:32:5)
#3      new Store (package:objectbox/src/native/store.dart:105:7)
#4      _MyHomePageState.initState.<anonymous closure> (package:untitled/main.dart:63:20)
#5      _rootRunUnary (dart:async/zone.dart:1362:47)
#6      _CustomZone.runUnary (dart:async/zone.dart:1265:19)
#7      _FutureListener.handleValue (dart:async/future_impl.dart:152:18)
#8      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:704:45)
#9      Future._propagateToListeners (dart:async/future_impl.dart:733:32)
#10     Future._completeWithValue (dart:async/future_impl.dart:539:5)
#11     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:254:13)
#12     getTemporaryDirectory (package:path_provider/path_provider.dart)
<asynchronous suspension>

Syncing files to device macOS...                                    88ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on macOS is available at: http://127.0.0.1:59804/kwcL1lLkKYA=/

Flutter DevTools, a Flutter debugger and profiler, on macOS is available at: http://127.0.0.1:9106?uri=http%3A%2F%2F127.0.0.1%3A59804%2FkwcL1lLkKYA%3D%2F

💪 Running with sound null safety 💪

```

```
import 'package:flutter/material.dart';
import 'package:objectbox/objectbox.dart';
import 'package:path_provider/path_provider.dart';

import 'objectbox.g.dart';

@Entity()
class Note {
  int id;

  String text;
  String? comment;
  DateTime date;

  Note(this.text, {this.id = 0, this.comment, DateTime? date})
      : date = date ?? DateTime.now();
}

void main() {
  runApp(MyHomePage());
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  void initState() {
    super.initState();
    getTemporaryDirectory().then((value) {
      var _store =
          Store(getObjectBoxModel(), directory: value.path + '/objectbox');
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container();
  }
}

```

```
flutter doctor -v
[✓] Flutter (Channel stable, 2.0.6, on macOS 11.3.1 20E241 darwin-arm, locale zh-Hans)
    • Flutter version 2.0.6 at /Users/jochen/Documents/dev/flutter
    • Framework revision 1d9032c7e1 (3 weeks ago), 2021-04-29 17:37:58 -0700
    • Engine revision 05e680e202
    • Dart version 2.12.3
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/jochen/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.56.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 11.3.1 20E241 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 90.0.4430.212
```

--- a/objectbox/README.md
+++ b/objectbox/README.md
@@ -69,6 +69,20 @@ dev_dependencies:
 
 * Install the packages: `flutter pub get`
 * XCode/iOS only: increase the deployment target to iOS 11 and, under Architectures, replace `${ARCHS_STANDARD}` with `arm64` (or `$ARCHS_STANDARD_64_BIT`). See [FAQ](#faq) for details.
+* If targeting macOS (creating sandboxed macOS apps): you need to specify an application group.
+  Update all `macos/Runner/*.entitlements` files, adding the following code inside the `<dict>` section. 
+  Change the string value to the `DEVELOPMENT_TEAM` you can find in your Xcode settings, plus an application-specific suffix, for example: 
+  
+  ```xml
+  <key>com.apple.security.application-groups</key>
+  <array>
+    <string>FGDTDLOBXDJ.demo</string>
+  </array>
+  ```
+  
+  Next, in your app code, pass the same string when opening the Store, for example: `Store(getObjectBoxModel(), directory: ..., macosApplicationGroup: 'FGDTDLOBXDJ.demo')`.  
+  Note: Pick a short group identifier; there's an internal limit in macOS that requires the complete string to be 
+  20 characters or fewer, including a trailing slash that's added automatically.
 * Sync + Android only: in your `android/app/build.gradle` set `minSdkVersion 21` in section `android -> defaultConfig`. 
 
 ### Dart Native

--- a/objectbox/README.md
+++ b/objectbox/README.md
@@ -70,7 +70,7 @@ dev_dependencies:
 * Install the packages: `flutter pub get`
 * XCode/iOS only: increase the deployment target to iOS 11 and, under Architectures, replace `${ARCHS_STANDARD}` with `arm64` (or `$ARCHS_STANDARD_64_BIT`). See [FAQ](#faq) for details.
 * If targeting macOS (creating sandboxed macOS apps): you need to specify an application group.
-  Update all `macos/Runner/*.entitlements` files, adding the following code inside the `<dict>` section. 
+  Check all `macos/Runner/*.entitlements` files if they contain a `<dict>` section with correct group ID info. 
   Change the string value to the `DEVELOPMENT_TEAM` you can find in your Xcode settings, plus an application-specific suffix, for example: 
   
   ```xml
@@ -82,7 +82,7 @@ dev_dependencies:
   
   Next, in your app code, pass the same string when opening the Store, for example: `Store(getObjectBoxModel(), directory: ..., macosApplicationGroup: 'FGDTDLOBXDJ.demo')`.  
   Note: Pick a short group identifier; there's an internal limit in macOS that requires the complete string to be 
-  20 characters or fewer, including a trailing slash that's added automatically.
+  19 characters or fewer.
 * Sync + Android only: in your `android/app/build.gradle` set `minSdkVersion 21` in section `android -> defaultConfig`. 
 
 ### Dart Native

--- a/objectbox/example/flutter/objectbox_demo/lib/main.dart
+++ b/objectbox/example/flutter/objectbox_demo/lib/main.dart
@@ -52,7 +52,8 @@ class ViewModel {
   ViewModel(Directory dir)
       : _store = Store(getObjectBoxModel(),
             directory: dir.path + '/objectbox',
-            macosApplicationGroup: 'objectbox.demo') {
+            macosApplicationGroup: 'objectbox.demo' // replace with a real name
+            ) {
     _box = Box<Note>(_store);
     final qBuilder = _box.query()..order(Note_.date, flags: Order.descending);
     _queryStream = qBuilder.watch(triggerImmediately: true);

--- a/objectbox/example/flutter/objectbox_demo/lib/main.dart
+++ b/objectbox/example/flutter/objectbox_demo/lib/main.dart
@@ -50,8 +50,9 @@ class ViewModel {
   late final Stream<Query<Note>> _queryStream;
 
   ViewModel(Directory dir)
-      : _store =
-            Store(getObjectBoxModel(), directory: dir.path + '/objectbox') {
+      : _store = Store(getObjectBoxModel(),
+            directory: dir.path + '/objectbox',
+            macosApplicationGroup: 'objectbox.demo') {
     _box = Box<Note>(_store);
     final qBuilder = _box.query()..order(Note_.date, flags: Order.descending);
     _queryStream = qBuilder.watch(triggerImmediately: true);

--- a/objectbox/example/flutter/objectbox_demo/macos/Runner/DebugProfile.entitlements
+++ b/objectbox/example/flutter/objectbox_demo/macos/Runner/DebugProfile.entitlements
@@ -8,5 +8,9 @@
 	<true/>
 	<key>com.apple.security.network.server</key>
 	<true/>
+  <key>com.apple.security.application-groups</key>
+  <array>
+    <string>objectbox.demo</string>
+  </array>
 </dict>
 </plist>

--- a/objectbox/example/flutter/objectbox_demo/macos/Runner/Release.entitlements
+++ b/objectbox/example/flutter/objectbox_demo/macos/Runner/Release.entitlements
@@ -4,5 +4,9 @@
 <dict>
 	<key>com.apple.security.app-sandbox</key>
 	<true/>
+  <key>com.apple.security.application-groups</key>
+  <array>
+    <string>objectbox.demo</string>
+  </array>
 </dict>
 </plist>

--- a/objectbox/example/flutter/objectbox_demo_sync/lib/main.dart
+++ b/objectbox/example/flutter/objectbox_demo_sync/lib/main.dart
@@ -52,7 +52,8 @@ class ViewModel {
 
   ViewModel(Directory dir)
       : _store = Store(getObjectBoxModel(),
-            directory: dir.path + '/objectbox-sync') {
+            directory: dir.path + '/objectbox',
+            macosApplicationGroup: 'objectbox.demo') {
     _box = Box<Note>(_store);
     final qBuilder = _box.query()..order(Note_.date, flags: Order.descending);
     _queryStream = qBuilder.watch(triggerImmediately: true);

--- a/objectbox/example/flutter/objectbox_demo_sync/lib/main.dart
+++ b/objectbox/example/flutter/objectbox_demo_sync/lib/main.dart
@@ -53,7 +53,8 @@ class ViewModel {
   ViewModel(Directory dir)
       : _store = Store(getObjectBoxModel(),
             directory: dir.path + '/objectbox',
-            macosApplicationGroup: 'objectbox.demo') {
+            macosApplicationGroup: 'objectbox.demo' // replace with a real name
+            ) {
     _box = Box<Note>(_store);
     final qBuilder = _box.query()..order(Note_.date, flags: Order.descending);
     _queryStream = qBuilder.watch(triggerImmediately: true);

--- a/objectbox/example/flutter/objectbox_demo_sync/macos/Runner/DebugProfile.entitlements
+++ b/objectbox/example/flutter/objectbox_demo_sync/macos/Runner/DebugProfile.entitlements
@@ -8,5 +8,9 @@
 	<true/>
 	<key>com.apple.security.network.server</key>
 	<true/>
+  <key>com.apple.security.application-groups</key>
+  <array>
+    <string>objectbox.demo</string>
+  </array>
 </dict>
 </plist>

--- a/objectbox/example/flutter/objectbox_demo_sync/macos/Runner/Release.entitlements
+++ b/objectbox/example/flutter/objectbox_demo_sync/macos/Runner/Release.entitlements
@@ -4,5 +4,9 @@
 <dict>
 	<key>com.apple.security.app-sandbox</key>
 	<true/>
+  <key>com.apple.security.application-groups</key>
+  <array>
+    <string>objectbox.demo</string>
+  </array>
 </dict>
 </plist>

--- a/objectbox/lib/src/native/store.dart
+++ b/objectbox/lib/src/native/store.dart
@@ -71,7 +71,8 @@ class Store {
       int? maxDBSizeInKB,
       int? fileMode,
       int? maxReaders,
-      bool queriesCaseSensitiveDefault = true})
+      bool queriesCaseSensitiveDefault = true,
+      String? macosApplicationGroup})
       : _weak = false,
         _queriesCaseSensitiveDefault = queriesCaseSensitiveDefault,
         _dbDir = path.context.canonicalize(
@@ -79,6 +80,21 @@ class Store {
                 ? 'objectbox'
                 : directory) {
     try {
+      if (Platform.isMacOS && macosApplicationGroup != null) {
+        if (!macosApplicationGroup.endsWith('/')) {
+          macosApplicationGroup += '/';
+        }
+        if (macosApplicationGroup.length > 20) {
+          ArgumentError.value(macosApplicationGroup, 'macosApplicationGroup',
+              'Must be at most 20 characters long');
+        }
+        final cStr = macosApplicationGroup.toNativeUtf8();
+        try {
+          C.posix_sem_prefix_set(cStr.cast());
+        } finally {
+          malloc.free(cStr);
+        }
+      }
       if (_openStoreDirectories.contains(_dbDir)) {
         throw UnsupportedError(
             'Cannot create multiple Store instances for the same directory. '

